### PR TITLE
Adds notes on logging

### PR DESCRIPTION
**Pull request by perrymant**
These are my notes on logging.

## Checklist for the author

- [x] I have done a self-review before requesting a review from someone else.
- [x] I believe the PR should not be broken down into smaller PRs as it regards only one piece of work.
- [x] I have provided a good title and description.
- [x] I have raised any questions in Slack instead of in the PR or in comments in code.
- [x] I have assigned a reviewer I believe has the background information needed to review this.

## Checklist for the reviewer

- [x] I have been nice to the author in my review.
- [x] I have been as clear as I can in my comments, and cannot think of a better way of formulating any of my comments.
- [x] I have been constructive with my comments. I have provided alternatives to issues I've spotted, instead of just pointing out that they're wrong. 
- [x] I have used inline comments when possible.
- [x] I have used 'Request changes' only for changes where the author might wish to state a reason for a certain approach, or if I believe I should have a second look to verify everything is OK. I have approved the PR otherwise, with the understanding that my comments will be addressed.
- [x] I have not been picky on personal code style when that's not the point of the PR
- [x] I have tested the change locally if there's non-trivial code involved.

### CODE REVIEW

File: students/perrymant/22-Logging.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    - `ERROR`: Error log messages are used to relay errors that occur at various levels in a computer system. Unfortunately, many error messages only give you a starting point as to why they occurred.
+    - `ALERT`: An alert is meant to indicate that something interesting has happened.
+
+- The levels for `java.util.logging` (the underlying loggin framework in Java) in descending order are:
+    - `SEVERE` (highest value): indicating a serious failure
+    - `WARNING`: indicating a potential problem
+    - `INFO`: a message level for informational messages.
+    - `CONFIG`: a message level for static configuration messages, to assist in debugging problems that may be associated with particular configurations.
+    - `FINE`, `FINER`, `FINEST`  (lowest value)
+    All of `FINE`, `FINER`, and `FINEST` are intended for relatively detailed tracing.  The exact meaning of the three levels will vary between subsystems, but in general, `FINEST` should be used for the most voluminous detailed output, `FINER` for somewhat less detailed output, and `FINE` for the  lowest volume (and most important) messages.
```
Comment (javarb) [inline students/perrymant/22-Logging.md]: As I said, I think this can be attached to the previous bullet point.

File: students/perrymant/22-Logging.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+- In Syslog, the logging framework for OS and Networking, the levels are:
```
Comment (javarb) [inline students/perrymant/22-Logging.md]:
```suggestion
- In Syslog, the logging framework for OS and Networking, the log levels are:
```

File: students/perrymant/22-Logging.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+| Alert         | Action must be taken immediately  | A condition that should be corrected immediately, such as a corrupted system database. |
+| Critical      | Critical conditions               | Hard device errors.                                                                    |
+| Error         | Error conditions                  |                                                                                        |
+| Warning       | Warning conditions                |                                                                                        |
+| Notice        | Normal but significant conditions | Conditions that are not error conditions, but that may require special handling.       |
+| Informational | Informational messages            |                                                                                        |
+| Debug         | Debug-level messages              | Messages that contain information normally of use only when debugging a program.       |
+(source [Wiki](https://en.wikipedia.org/wiki/Syslog))
```
Comment (javarb) [inline students/perrymant/22-Logging.md]: I think you need a space here because is showed on a new row. Or you could put above so:

- In Syslog, the logging framework for OS and Networking, the log levels are (source [wiki](http://...)):

File: students/perrymant/22-Logging.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+- In Spring Boot Logging Levels are: `OFF`, `ERROR`, `WARN`, `INFO`, `DEBUG`, `TRACE`.
+- By default, Spring Boot configures logging via [Logback](http://logback.qos.ch) to write to the console at an INFO level.
```
Comment (javarb) [inline students/perrymant/22-Logging.md]:
```suggestion
- By default, Spring Boot configures logging via [Logback](http://logback.qos.ch) to write to the console at an `INFO` level.
```

File: students/perrymant/22-Logging.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+### Common log libraries in Java
+- `SLF4J` (Simple Logging Facade for Java, a facade is an object that serves as a front-facing interface masking more complex underlying or structural code. Here's a `HelloWorld` using `SLF4J`:
```
Comment (javarb) [inline students/perrymant/22-Logging.md]:
```suggestion
- `SLF4J` (Simple Logging Facade for Java). A facade is an object that serves as a front-facing interface masking more complex underlying or structural code. Here's the `HelloWorld` program using `SLF4J`:
```

File: students/perrymant/22-Logging.md [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+import org.slf4j.LoggerFactory;
+
+public class HelloWorld {
+  public static void main(String[] args) {
+    Logger logger = LoggerFactory.getLogger(HelloWorld.class);
+    logger.info("Hello World");
+  }
+}
+```
+This would result in the following log message: `0 [main] INFO HelloWorld - Hello World`
```
Comment (javarb) [inline students/perrymant/22-Logging.md]:
```suggestion
Once this program is executed, the following log message is shown: `0 [main] INFO HelloWorld - Hello World`
```

File: students/perrymant/22-Logging.md [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+- Here's an example with `log4j`, logging in 3 different log levels:
```
Comment (javarb) [inline students/perrymant/22-Logging.md]:
```suggestion
- `Log4j` is another library for logging in Java. Here's a log example setting a log message in three different log levels using `Log4J`:
```

File: students/perrymant/22-Logging.md [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    private static Logger logger = LogManager.getLogger(Log4jExample.class);
+
+    public static void main(String[] args) {
+        logger.debug("Debug log message");
+        logger.info("Info log message");
+        logger.error("Error log message");
+    }
+}
+```
+This would result in the following log message:
```
Comment (javarb) [inline students/perrymant/22-Logging.md]:
```suggestion
Once this program is executed, the following log messages are shown: 
```

File: students/perrymant/22-Logging.md [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+- A log like `May 21 14:38:25 error` would not be very meaningful, so we should consider improving the quality of the logs that our applications generate, including as much useful information for our logs to help with deriving meaning.
+- The typical basic contents for a log message are the following:
+    - Timestamp
+    - Source
+    - Data
+- In general, logs should tell you:
+    - What happened (with appropriate detail; “Something happened” is not usually particularly useful).
+    - When did it happen (and when did it start and end, if relevant).
+    - Where did it happen (on what host, what file system, which network interface, etc.).
+    - Who was involved.
```
[the remark this answers is not in the document]
Comment (kinbiko) [reply students/perrymant/22-Logging.md]: Good point. I think perhaps a user ID is even more common -- it's usually less sensitive information, but still uniquely identifies the right entity.

File: students/perrymant/22-Logging.md [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+    - Where he, she or it came from (if dealing with OS/Security logs).
+- The logging ecosystem (or logging infrastructure), are all the components and piece parts that come together to allow for the generation, filtering, normalization, analysis, and long-term storage of log data.
+    Once a device or computer system is configured to generate log messages, the next step is to filter and normalize the messages.
+    - `Filtering` deals with including or excluding log messages based on the content in the log message.
+    - `Normalization` is the act of taking disparately formatted log messages and converting them to a common format through the use of parsing, which entails scanning the log message from start to finish to pull out information we are interested in and place them into normalized fields.
+    - `Log message analysis` deals with collating, correlating and analyzing log data to derive meaning from it.
+    - `long-term data store` is typically accomplished by using a database.
```
[the remark this answers is not in the document]
Comment (kinbiko) [reply students/perrymant/22-Logging.md]: Yeah, I think @javarb is mostly right about this one.
Although! It is pretty common to store your logs in Elasticsearch (a database), and use something called the ELK stack (Elasticsearch, Logstash, Kibana) to centralise your logs.

There are entire companies built around solving the problem @javarb mentions.

File: students/perrymant/22-Logging.md [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+}
+```
+This would result in the following log message:
+```
+15:11:22.022 [main] DEBUG Log4jExample - Debug log message
+15:11:22.027 [main] INFO Log4jExample - Info log message
+15:11:22.027 [main] ERROR Log4jExample - Error log message
+```
+
+For full control over the logging configuration in Intellij, you can create a logback.xml file at the root of the classpath (in src/main/resources). Here’s an example of a simple log-back.xml file you might use (this Logback configuration is more or less equivalent to the default you’ll get if you have no logback.xml file.):
```
[the remark this answers is not in the document]
Comment (kinbiko) [reply students/perrymant/22-Logging.md]: logback.xml or log-back.xml? I'm guessing it's the former.

File: students/perrymant/22-Logging.md [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+- Typical log levels:
+    - `INFORMATIONAL`: Messages of this type are designed to let users and administrators know that something benign has occurred.
+    - `DEBUG`: Debug messages are generally generated from software systems in order to aid software developers troubleshoot and identify problems with running application code.
+    - `WARNING`: Warning messages are concerned with situations where things may be missing or needed for a system, but the absence of which will not impact system operation. For example, if a program isn’t given the proper number of command line arguments, but yet it can run without them, is something the program might log just as a warning to the user or operator.
+    - `ERROR`: Error log messages are used to relay errors that occur at various levels in a computer system. Unfortunately, many error messages only give you a starting point as to why they occurred.
+    - `ALERT`: An alert is meant to indicate that something interesting has happened.
+
+- The levels for `java.util.logging` (the underlying loggin framework in Java) in descending order are:
```
Comment (kinbiko) [inline students/perrymant/22-Logging.md]: Nobody really uses this framework so feel free to not learn about this.

File: students/perrymant/22-Logging.md (hunk 11 above)
Comment (kinbiko) [inline students/perrymant/22-Logging.md]: This is unrelated to IntelliJ. It's also not really the root of the class path, so I'd just say `src/main/resources` and don't mention the classpath.

File: students/perrymant/22-Logging.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (javarb) [reply students/perrymant/22-Logging.md]: I see @kinbiko, I think that completes the landscape about how to store logs currently and provides the workaround to the problem, thanks for the information!
